python/ray/util/ml_utils/checkpoint.py

```diff
@@ -0,0 +1,413 @@
+        local_checkpoint = LocalStorageCheckpoint(path=path, metadata=metadata)
+        return local_checkpoint
+
+    def to_object_ref(self) -> ray.ObjectRef:
+        return ray.put(self.to_dict())
+
+    @classmethod
+    def from_object_ref(cls, obj_ref: ray.ObjectRef) -> "ObjectStoreCheckpoint":
+        return ObjectStoreCheckpoint(obj_ref=obj_ref, metadata=None)
+
+    def to_uri(self, location: str) -> str:
+        path = self.to_directory(path=None)
+        assert is_cloud_target(location)
+        upload_to_bucket(bucket=location, local_path=path)
+        if not isinstance(self, FSStorageCheckpoint):
+            # Only delete directory if it was a temporary directory
+            shutil.rmtree(path)
+        return location
+
+    @classmethod
+    def from_uri(cls, location: str) -> "CloudStorageCheckpoint":
+        return CloudStorageCheckpoint(location=location)
+
+    def to_data_checkpoint(self) -> "DataCheckpoint":
+        raise NotImplementedError
+
+    def to_object_store_checkpoint(self) -> "ObjectStoreCheckpoint":
+        raise NotImplementedError
+
+    def to_local_storage_checkpoint(
+        self, path: Optional[str] = None
+    ) -> "LocalStorageCheckpoint":
+        raise NotImplementedError
+
+    def to_cloud_storage_checkpoint(self, location: str) -> "CloudStorageCheckpoint":
+        location = self.to_uri(location=location)
+        return CloudStorageCheckpoint(location=location)
+
+
+class DataCheckpoint(TuneCheckpoint, DataArtifact):
+    def __init__(self, data: Any, metadata: Any = None):
+        TuneCheckpoint.__init__(self, metadata=metadata)
+        DataArtifact.__init__(self, data=data)
+
+    @property
+    def is_fs_checkpoint(self):
+        return self.metadata.get("is_fs_checkpoint", False)
+
+    def to_dict(self) -> dict:
+        if isinstance(self.data, dict):
+            return self.data
+        else:
+            return {
+                "data": self.data,
+                "metadata": {**self.metadata, "is_data_checkpoint": True},
+            }
+
+    def to_directory(self, path: Optional[str] = None) -> str:
+        path = path if path is not None else temporary_checkpoint_dir()
+
+        new_metadata = self.metadata.copy()
+
+        os.makedirs(path, exist_ok=True)
+        # Drop marker
+        open(os.path.join(path, ".is_checkpoint"), "a").close()
+        if self.is_fs_checkpoint:
+            # Recover FS checkpoint from data
+            _unpack(self.data["data"], path)
+            new_metadata["is_data_checkpoint"] = False
+        else:
+            # Dump into checkpoint.pkl
+            checkpoint_data_path = os.path.join(path, "checkpoint.pkl")
+            with open(checkpoint_data_path, "wb") as f:
+                pickle.dump(self.data, f)
+            new_metadata["is_data_checkpoint"] = True
+
+        write_metadata(path, new_metadata)
+        return path
+
+    def to_data_checkpoint(self) -> "DataCheckpoint":
+        return self
```

Comment:
Should this return `self` or a copy of `self`? Every other `to_X` method returns a separate object

Reply:
The other methods also return `self` if needed. It's a utility to make sure checkpoints are in the right format, so I wouldn't create an object copy here. If people want to copy, they can use copy.copy() instead.